src/mds/Server.cc

```diff
@@ -4106,6 +4106,17 @@ void Server::handle_client_getattr(const MDRequestRef& mdr, bool is_lookup)
     if (r > 0)
       return; // delayed
 
+    // Do not batch if any xlock is held
+    if (!r) {
+      CInode *in = mdr->in[0];
+      if (((mask & CEPH_CAP_LINK_SHARED) && (in->linklock.is_xlocked())) ||
+          ((mask & CEPH_CAP_AUTH_SHARED) && (in->authlock.is_xlocked())) ||
+          ((mask & CEPH_CAP_XATTR_SHARED) && (in->xattrlock.is_xlocked())) ||
+          ((mask & CEPH_CAP_FILE_SHARED) && (in->filelock.is_xlocked()))) {
```

Comment:
We can make a lesser impact by making sure that the xlocker is never batched. This will allow batching multiple non-xlocker clients, which should be safe.

```suggestion
      if (((mask & CEPH_CAP_LINK_SHARED) && (in->linklock.is_xlocked_by_client(client))) ||
          ((mask & CEPH_CAP_AUTH_SHARED) && (in->authlock.is_xlocked_by_client(client))) ||
          ((mask & CEPH_CAP_XATTR_SHARED) && (in->xattrlock.is_xlocked_by_client(client))) ||
          ((mask & CEPH_CAP_FILE_SHARED) && (in->filelock.is_xlocked_by_client(client)))) {
```

Reply:
Yeah, this will be better.

Reply:
Fixed it and thanks @leonid-s-usov @gregsfortytwo

Reply:
Wait, please double check, my patch above is probably wrong, but the idea stays: don't batch when the xlocker client is the head.

Reply:
This is fine. All the xlocker client requests won't be batched and even won't be set as batch heads.

With this change all the non-locker client requests could be batched and the batch heads will acquire the `rdlock` and then wait.